sdk/resourcemanagerhybrid/azure-resourcemanager/src/main/java/com/azure/resourcemanager/AzureResourceManager.java

```diff
@@ -313,6 +324,57 @@ public AzureResourceManager withDefaultSubscription() {
         }
     }
 
+    /**
+     * Gets AzureEnvironment from ARM endpoint.
+     *
+     * @param armEndpoint the URL of ARM endpoint.
+     * @param httpClient the HTTP client for requests.
+     * @return the AzureEnvironment
+     */
+    public static AzureEnvironment getAzureEnvironmentFromArmEndpoint(String armEndpoint, HttpClient httpClient) {
+        HttpRequest request = new HttpRequest(HttpMethod.GET,
+            String.format("%s/metadata/endpoints?api-version=1.0", armEndpoint))
+            .setHeader("accept", "application/json");
+
+        HttpResponse response = httpClient.send(request).block();
+        if (response.getStatusCode() != 200) {
+            throw new HttpResponseException("Failed : HTTP error code : " + response.getStatusCode(), response);
+        }
+        String body = response.getBodyAsString().block();
+        try {
+            ArmMetadata metadata = JacksonAdapter.createDefaultSerializerAdapter()
+                .deserialize(body, ArmMetadata.class, SerializerEncoding.JSON);
+
+            AzureEnvironment azureEnvironment = new AzureEnvironment(new HashMap<>() {
+                private static final long serialVersionUID = 1L;
+                {
+                    put("managementEndpointUrl", metadata.authentication.audiences.get(0));
+                    put("resourceManagerEndpointUrl", armEndpoint);
+                    put("galleryEndpointUrl", metadata.galleryEndpoint);
+                    put("activeDirectoryEndpointUrl", metadata.authentication.loginEndpoint);
+                    put("activeDirectoryResourceId", metadata.authentication.audiences.get(0));
+                    put("activeDirectoryGraphResourceId", metadata.graphEndpoint);
+                    put("storageEndpointSuffix", armEndpoint.substring(armEndpoint.indexOf('.')));
+                    put("keyVaultDnsSuffix", ".vault" + armEndpoint.substring(armEndpoint.indexOf('.')));
```

Comment:
I copied from stack sample https://github.com/abbhar/azure-stack-java-samples/blob/d3a9015bf5e53eed13daaec9c17340ad6f248e44/samples/src/main/java/com/azure/resourcemanager/hybrid/samples/ManageVirtualMachine.java#L174

However I am not sure whether this 2 endpoint is correct.

Reply:
They are correct for Azure Stack only, but with a newer metadata API version (`api-version=2019-10-01`) they can be fetched directly from the response instead. There's also separate suffixes for user and admin keyvault depending on which ARM endpoint is used on the device.

The schema is similar except that everything for the environment is in a one-element array, and there's some endpoint property renames.  Can see an example of it with https://management.azure.com/metadata/endpoints?api-version=2019-10-01. 

```java
new AzureEnvironment(new HashMap<String, String>() {{
    put("managementEndpointUrl", metadata.get(0).authentication.audiences.get(0));
    put("resourceManagerEndpointUrl", armEndpoint);
    put("galleryEndpointUrl", metadata.get(0).gallery);
    put("activeDirectoryEndpointUrl", metadata.get(0).authentication.loginEndpoint);
    put("activeDirectoryResourceId", metadata.get(0).authentication.audiences.get(0));
    put("activeDirectoryGraphResourceId", metadata.get(0).graph);
    put("storageEndpointSuffix", "." + metadata.get(0).suffixes.storage);
    put("keyVaultDnsSuffix", "." + metadata.get(0).suffixes.keyVaultDns);
}})
```

Reply:
here is an azurestack endpoint that you have access to (on corpnet)
https://management.redmond.azurestack.corp.microsoft.com//metadata/endpoints?api-version=2019-10-01

https://adminmanagement.redmond.azurestack.corp.microsoft.com//metadata/endpoints?api-version=2019-10-01

We should use the 2019-10-01

Reply:
Changed to `api-version=2019-10-01`